--- a/sound/soc/ralink/ralink-i2s.c
+++ b/sound/soc/ralink/ralink-i2s.c
@@ -428,7 +428,14 @@ static int ralink_i2s_trigger(struct snd
 	case SNDRV_PCM_TRIGGER_STOP:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		val = 0;
+		/* Flush hardware TX/RX FIFO buffer before stopping clock */
+		val = ralink_i2s_read(priv, RALINK_I2S_CON);
+		val |= I2S_FIFO_CLR;
+		ralink_i2s_write(priv, RALINK_I2S_CON, val);
+		udelay(100);
+		val = 0;
 		break;
 	default:
 		return -EINVAL;
